Detectors/TPC/reconstruction/include/TPCReconstruction/RawReader.h

```diff
@@ -0,0 +1,319 @@
+
+    /// Get data
+    /// @param padPos local pad position (row starts with 0 in each region)
+    /// @return shared pointer to data vector, each element is one timebin
+    std::shared_ptr<std::vector<uint16_t>> getData(const PadPos& padPos);
+
+    /// Get data of next pad position
+    /// @param padPos local pad position (row starts with 0 in each region)
+    /// @return shared pointer to data vector, each element is one timebin
+    std::shared_ptr<std::vector<uint16_t>> getNextData(PadPos& padPos);
+
+    int getRegion() const { return mRegion; };
+    int getLink() const { return mLink; };
+    int getEventNumber() const { return mLastEvent; };
+    int getRunNumber() const { return mRun; };
+
+    void setUseRawInMode3(bool val) { mUseRawInMode3 = val; };
+    void setApplyChannelMask(bool val) { mApplyChannelMask = val; };
+    void setChannelMask(std::shared_ptr<CalDet<bool>> channelMask) { mChannelMask = channelMask; };
+    void setPrintRawData(bool val) { mPrintRawData = val; };
+    void setCheckAdcClock(bool val) { mCheckAdcClock = val; };
+
+    /// Returns some innformation about the event, e.g. the header
+    /// @param event Event number
+    /// @return shared pointer to vector with event informations
+    std::shared_ptr<std::vector<EventInfo>> getEventInfo(uint64_t event) const;
+
+    std::vector<std::tuple<short,short,short>> getAdcError() { return mAdcError; };
+
+  private:
+
+    bool decodeRawGBTFrames(EventInfo eventInfo);
+    bool decodePreprocessedData(EventInfo eventInfo);
+
+    int mRegion;                        ///< Region of the data
+    int mLink;                          ///< FEC of the data
+    int mRun;                           ///< Run number
+    bool mUseRawInMode3;                ///< in readout mode 3 decode GBT frames
+    bool mApplyChannelMask;             ///< apply channel mask
+    bool mCheckAdcClock;                ///< check the ADC clock
+    bool mPrintRawData;                 ///< print the RAW data while decoding
+    int64_t mLastEvent;                 ///< Number of last loaded event
+    std::array<uint64_t,5> mTimestampOfFirstData;   ///< Time stamp of first decoded ADC value, individually for each half sampa
+    std::map<uint64_t, std::shared_ptr<std::vector<EventInfo>>> mEvents;                ///< all "event data" - headers, file path, etc. NOT actual data
+    std::map<PadPos,std::shared_ptr<std::vector<uint16_t>>> mData;                      ///< ADC values of last loaded Event
+    std::map<PadPos,std::shared_ptr<std::vector<uint16_t>>>::iterator mDataIterator;    ///< Iterator to last requested data
+    std::array<short,5> mSyncPos;       ///< positions of the sync pattern (for readout mode 3)
+
+    std::shared_ptr<CalDet<bool>> mChannelMask;     ///< Channel mask
+    std::vector<std::tuple<short,short,short>> mAdcError;
+};
+
+inline
+std::shared_ptr<std::vector<RawReader::EventInfo>> RawReader::getEventInfo(uint64_t event) const {
```

Comment:
do you really need `std::shared_ptr` (as in a reference counted thing)? The alternative is using `std::uniq_ptr` for the ownership and passing arround bare pointers (might not apply here).